include/audio/dmic.h

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright (c) 2018, Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author:	Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+ *		Sathish Kuttan <sathish.k.kuttan@.intel.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef __AUDIO_DMIC_H__
+#define __AUDIO_DMIC_H__
+
+#include <kernel.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * DMIC driver states
+ */
+enum dmic_state {
+	DMIC_STATE_UNINIT,	/* Uninitialized */
+	DMIC_STATE_INITIALIZED,	/* Initialized */
+	DMIC_STATE_CONFIGURED,	/* Configured */
+	DMIC_STATE_ACTIVE,	/* Active */
+	DMIC_STATE_PAUSED,	/* Paused */
+};
+
+/**
+ * DMIC driver trigger commands
+ */
+enum dmic_trigger {
+	DMIC_TRIGGER_STOP,	/* stop stream */
+	DMIC_TRIGGER_START,	/* start stream */
+	DMIC_TRIGGER_PAUSE,	/* pause the stream */
+	DMIC_TRIGGER_RELEASE,	/* release paused stream */
+	DMIC_TRIGGER_RESET,	/* reset */
+};
+
+/**
+ * PDM Channels LEFT / RIGHT
+ */
+enum pdm_lr {
+	PDM_CHAN_LEFT,
+	PDM_CHAN_RIGHT,
+};
+
+/**
+ * PDM Input/Output signal configuration
+ */
+struct pdm_io_cfg {
+	/* parameters global to all PDM controllers */
+
+	/* minimum clock frequency supported by the mic */
+	u32_t	min_pdm_clk_freq;
```

Comment:
How clock frequency and duty cycle is determined? Could we just have per-mic/per-controller configuration?

Reply:
They are specified in the datasheet of the microphones.
We could have per-mic, but it is unnecessary. Typically all microphones in an array are from the same vendor and same type. It's an overkill to define the API to allow this to be specified for each mic individually especially if the mic count is high.